[DONOTMERGE] (Pending 5169) OpenShift Volume Security Integration

pweil- opened the pull request:
This is based on @pmorie 's WIP branch https://github.com/openshift/origin/pull/5169.  The last two commits are what I'm working on.

Contains:

1.  SecurityContextConstraints api support
   1.  Includes a defaulter for SCCs without FSGroup/SupplementalGroups strategies so no upgrade is necessary
   1.  Defaulting is based on the value of the RunAsUser strategy
1.  Strategies for groups (fsgroup and sup groups use the same underlying strategy code but have their own definitions/types)
   1.  Both strategies allow namespace configuration by using MustRunAs with no ranges set
   1.  Both strategies will provide a default value of the first value in the first range configured
   1.  When using namespace configuration for the FSGroup strategy it will create a single range with a min/max of the first value in the supplemental group annotation. 
   1.  When using namespace configuration for the SupGroup strategy it will parse the entire annotation into ranges.
1.  provider implementation
   1.  new interface methods introduced to generate and validate a PSC 
   1.  support for fsgroup/sup group strategies
   1.  reuse of the selinux strategy validation on the PSC by mocking up a container to validate (generation never uses the container so no mock is necessary)
1.  admission control for PSC
   1.  generates and validates the PSC
   1.  uses the SecurityContext provider to determine the effective SC for the container prior to container generation/validation (provides us initial value defaulting and allows container overrides)
   1.  retrieval of the supplemental group annotation has fallback code to use the uid annotation if necessary for a better upgrade experience

@pmorie @swagiaal @liggitt @smarterclayton 

---
Remaining Work:
 - [ ] blacklisting `v1beta3` `PodSpecs` has rippled into a lot of objects now not getting fuzzed.  At the very least, we need an issue find a way to test `v1beta3 -> internal -> v1beta3` with fidelity.  See the ripples here: https://github.com/openshift/origin/pull/5079/files#diff-394377845c8dfd9eec363c059cdde354R379
 - [ ] UPSTREAM commit names need fixing.
 - [ ] issue to allow admission based on a union of available SCCs.  With multiple ranges and options, its really annoying to have an SCC that combines all of them.
 - [ ] issue to convert code to a `DeepCopy` of the pod.  This would help readability and its only going to happen on pod creation.  I'm not convinced that an in-memory copy of data we already have, just for creates of pods, will show up on profiling, but it be a big improvement for inspectability of this code.
 - [ ] issue for `reconcile-sccs` to help admins update correctly.

pmorie commented:
@pweil- I think we should make ginkgo tests for origin that test the knobs for docker in conjunction with the policy.  Long-term we need to have that for everything, we can start with just the scenarios being added (IE, SupplementalGroups and FSGroup).  Thoughts?

pweil- commented:
are you proposing carrying upstream ginko tests that work with our carried scc or just adding them to security policy when it goes in the exp api?

pmorie commented:
@pweil- Short term, i'm proposing e2es for origin that are probably similar to the ones for the knobs we make upstream that work in concert w/ the origin policy but don't share code with the upstream E2Es.  Ideally we would be able to intercept and decorate namespace creation in origin to set up policy but I don't want to have to work through that to solve the problem of just adding E2E for this feature in origin.

pweil- commented:
latest update has supplemental groups in providers and strategies along with tests (still just the last two commits)

pweil- commented:
provider using restricted scc to default the supplemental group to the first value of the allowable range

```
"securityContext": {
            "supplementalGroups": [
                1000010000
            ]
        },

"GroupAdd": [
            "1000010000"
        ],
```

pweil- commented:
@smarterclayton @pmorie @swagiaal @liggitt - fsgroup integration is in for SCC.  The last two commits in this branch are what needs eyes, the rest are changes from @pmorie and @swagiaal that will be reviewed and submitted separately. 

```
"securityContext": {
            "supplementalGroups": [
                1000010000
            ],
            "fsGroup": 1000010000
        },
```

pweil- commented:
@swagiaal  @pmorie  - a couple of things I'm noticing:

1.  There is nothing accounting for the new PodSecurityContext in the OpenShift v1beta3 code so round tripping a v1beta3 object would lose the PSC (make check is failing because the new fields aren't retained)
1.  swagger needs to be updated, when I verify on my branch it's showing changes for some of these commits
1.  gofmt on the commits is failing

pweil- commented:
took care of the feedback for making constant field names in the provider.  Awaiting other reviews.  

reviewers: please pay attention to the conversions.  Since the upstream has diverged (it is now PSC) there is not a nice way to auto generate these so a double check would be good.

smarterclayton commented:
Add one integration test that validates that the default behavior is to allocate the group id into the container

pweil- commented:
added v1/v1beta3 defaults if new group strategy types are empty. Defaults will be for a restricted, namespace allocated strategy (mustRunAs with no ranges set).  

added integration test that submits pods with and without requested groups and ensures the groups are allocated if necessary and that they end up on the docker host config.

pweil- commented:
updated to latest from @pmorie.  This now contains all required pieces.  Related PR that this depends on https://github.com/openshift/origin/pull/5169.

pweil- commented:
@liggitt @pmorie @smarterclayton @deads2k - please put this in your review queue for the last 2 commits.  They are ready to wrap up.  Summary is in the main comment.

@deads2k - question on the extensions conversions/deep copies.  I notice in there that there is not a conversion/copy for things like Node and Pod.  But there is a conversion for PodSpec.  Do you know why?  I did not add conversions for SCC to extensions but I have them in another PR if they are necessary.

pweil- commented:
@soltysh too for reviews

pweil- commented:
Bump.  Monday mornings are perfect for SCC reviews.  Best to do when you wake up angry :smile:

deads2k commented:
> @deads2k - question on the extensions conversions/deep copies. I notice in there that there is not a conversion/copy for things like Node and Pod. But there is a conversion for PodSpec. Do you know why? I did not add conversions for SCC to extensions but I have them in another PR if they are necessary.

the `PodSpec` was transitively required through the `DaemonSetSpec`.  If your object lives under an object that has conversions here, you probably need it.  The really scary question to ask yourself is why they have generated the same conversion function twice?  I suspect it was done because they did not properly indicate which schemes/group/version a given group/version relies upon.

pweil- commented:
> the PodSpec was transitively required through the DaemonSetSpec. If your object lives under an object that has conversions here, you probably need it. The really scary question to ask yourself is why they have generated the same conversion function twice? I suspect it was done because they did not properly indicate which schemes/group/version a given group/version relies upon.

Ok, should be safe then as SCC is a standalone object.

deads2k commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -41,45 +51,81 @@ func NewSimpleProvider(scc *api.SecurityContextConstraints) (SecurityContextCons
 	}, nil
 }
 
+// Create a PodSecurityContext based on the given constraints.  If a settingis already set
```
setting is

deads2k commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -41,45 +51,81 @@ func NewSimpleProvider(scc *api.SecurityContextConstraints) (SecurityContextCons
 
+// Create a PodSecurityContext based on the given constraints.  If a settingis already set
+// on the PodSecurityContext it will not be changed.  Validate should be used after the context
+// is created to ensureit complies with the required restrictions.
```
ensure it

deads2k commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -41,45 +51,81 @@ func NewSimpleProvider(scc *api.SecurityContextConstraints) (SecurityContextCons
+	}
+
+	if sc.FSGroup == nil {
+		fsGroup, err := s.fsGroupStrategy.Generate(pod)
```
Why does this return a list if we're only using the first value?

pweil- replied:
I want to be able to reuse the same scc strategies for sup group and fsgroup.  This allows custom strategies to provide ranges of acceptable groups and multiple defaults.  I thought it best to leave the interface flexible enough to provide multiple values.

Two things to note:

1.  fsgroup is a single field (so to reuse I need to make sure I only take a single value)
2.  the default mustRunAs generation is just returning the first value in the configured ranges anyway (but as mentioned, I thought it best to leave the interface capable of returning multiple to support defaulting to multiple sup groups)

deads2k replied:
> the default mustRunAs generation is just returning the first value in the configured ranges anyway (but as mentioned, I thought it best to leave the interface capable of returning multiple to support defaulting to multiple sup groups)

Docs on the interface usage describe this behavior?  I think you shoudl still have two different interfaces, even if they overlap.

pweil- replied:
I can live with that.  I'll add a generate single method.

deads2k commented on pkg/api/serialization_test.go:
```diff
@@ -375,7 +375,9 @@ func roundTrip(t *testing.T, codec runtime.Codec, originalItem runtime.Object) {
 	// docker's registry API.
-	"DockerImage": {"pre012", "1.0"},
+	"DockerImage":          {"pre012", "1.0"},
+	"DeploymentConfigList": {"", "v1"},
```
This is because we're breaking compatibility with `v1beta3`?  Where are the special cased tests to make sure I can still correctly parse these from `v1beta3` in?

pweil- replied:
@pmorie :arrow_up:

deads2k commented on UPGRADE.md:
```diff
@@ -32,8 +32,26 @@ At that time, the openshift docker registry image must be upgraded in order to c
+  which will set this field to true for the privileged SCC and false for the restricted SCC.
+  1.  fsGroup - if the strategy type is unset this field will default based on the runAsUser strategy.
+  If runAsUser is set to RunAsAny this field will also be set to RunAsAny.  If the strategy type is
+  any other value this field will default to MustRunAs and look to the namespace for annotation 
```
And when that annotation is missing, what will it do?

deads2k commented on pkg/api/serialization_test.go:
```diff
@@ -385,7 +387,7 @@ func TestSpecificKind(t *testing.T) {
 	defer api.Scheme.Log(nil)
 
-	kind := "ImageStreamTag"
+	kind := "BuildConfig"
```
This is a fun record of who had trouble with what most recently :)

deads2k commented on pkg/security/util.go:
```diff
@@ -3,8 +3,7 @@ package security
-	// for the namespace.  Groups are in the form of individual group ids or a range of group ids.
-	// Range format is supported in the form of a Block which supports {start}/{length} or {start}-{end}
+	// for the namespace.  Groups are in the form of a Block which supports {start}/{length} or {start}-{end}
 	SupplementalGroupsAnnotation = "openshift.io/sa.scc.supplemental-groups"
```
`sa`?

pweil- replied:
Yeah, I kept the new name consistent with the old names which have sa in there.

deads2k replied:
> Yeah, I kept the new name consistent with the old names which have sa in there.

Ok.  For the historical record, what does that stand for?

pweil- replied:
It was originally meant to stand for service account but is not applicable since an SA isn't directly assigned these values, the ns is.  I will remove it from the newest annotation to not perpetuate wrongness for the sake of consistency unless you disagree with that approach and would rather keep it consistent.

deads2k replied:
> It was originally meant to stand for service account but is not applicable since an SA isn't directly assigned these values, the ns is. I will remove it from the newest annotation to not perpetuate wrongness for the sake of consistency unless you disagree with that approach and would rather keep it consistent.

Go ahead and keep them consistent.

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -177,11 +196,13 @@ func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod
-	// pod so let's apply what we generated
+	// pod so let's apply what we generated.  Note: the psc is already applied.
 	for i, sc := range generatedSCs {
 		pod.Spec.Containers[i].SecurityContext = sc
```
Haven't I already done this here: https://github.com/openshift/origin/pull/5079/files#diff-eb43ba779f411e3ee6b772cceee9810dR194

pweil- replied:
Think you worked this one out yourself below.  Making the naming change to the loop above so it is more obvious that we aren't changing the original container.

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -177,11 +196,13 @@ func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod
 	// if we've reached this code then we've generated and validated an SC for every container in the
-	// pod so let's apply what we generated
+	// pod so let's apply what we generated.  Note: the psc is already applied.
 	for i, sc := range generatedSCs {
```
Why don't I just iterate over the `pod.Spec.Containers`?  That would be more obvious to me as a reader.

deads2k replied:
> Why don't I just iterate over the pod.Spec.Containers? That would be more obvious to me as a reader.

I think I see now.

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -164,7 +165,25 @@ func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod
+	pod.Spec.SecurityContext = psc
+	errs = append(errs, provider.ValidatePodSecurityContext(pod).Prefix("spec.securityContext")...)
+
 	for i, c := range pod.Spec.Containers {
```
Ok, looking through this loop, its really hard to read what's modifying a container and what is modifying a copy of a container that is going to be discarded later.

Re-reading, I **think** that this use a pointer to a copy of a container to perform a bunch of mutations that are ultimately discarded.  If that's right, can you rename `c` to `containerCopy` or some such?

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -216,40 +240,51 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
 				continue
 			}
 
```
Why can't I close the `if requiresNamespaceAllocations` here?

pweil- replied:
resolving the ns values is only applicable to something that needs ns allocation.  It could be closed since the other flags gate resolving individual values if that helps readability.

deads2k replied:
> resolving the ns values is only applicable to something that needs ns allocation. It could be closed since the other flags gate resolving individual values if that helps readability.

Yes, close here.  It helps break this apart into congruent chunks with a precondition.

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -216,40 +240,51 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
 
+			// Make a copy of the constraint so we don't mutate the store's cache
+			var constraintCopy kapi.SecurityContextConstraints = *constraint
+			constraint = &constraintCopy
```
This is warping my mind a little.  Can we move it to the top of the loop, choose a non-conflicting name, and assign the final thing to `constraintCopy`?  Something like this:

```go
for _, scc := range sccs{
    var t kapi.SecurityContextConstraints = *scc
    sccCopy := &t
```

The name change would ensure you don't accidentally leave any stragglers.

deads2k replied:
Why not make this a deep copy to avoid these shenanigans later on: https://github.com/openshift/origin/pull/5079/files#diff-eb43ba779f411e3ee6b772cceee9810dR262

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -205,8 +226,11 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
 
-		if resolveUIDRange || resolveSELinuxLevel {
+		if requiresNamespaceAllocations {
 			var min, max *int64
```
seems like these should be scoped to the blocks where you need them, so that they never escape.

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -216,40 +240,51 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
+
 			// Resolve the values from the namespace
 			if resolveUIDRange {
 				if min, max, err = getPreallocatedUIDRange(namespace); err != nil {
```
shouldn't this be able to write as 
```go
constraint.RunAsUser.UIDRangeMin, constraint.RunAsUser.UIDRangeMax, err = getPreallocatedUIDRange(namespace)
if err != nil{
    // do stuff
}
```

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -349,6 +384,91 @@ func getPreallocatedLevel(ns *kapi.Namespace) (string, error) {
+	glog.V(4).Infof("got preallocated value for groups: %s in namespace %s", groups, ns.Name)
+
+	blocks, err := parseSupplementalGroupAnnotation(groups)
+	if err != nil {
```
Why aren't I allowed to have zero supplemental groups?

pweil- replied:
When using a strategy that requires values you'll get an error creating the strategy if no values are given.  This allows us to indicate that the error was in parsing the annotation rather than a generic "bad strategy" message.

deads2k commented:
I found some of the code hard to parse.  After its updated, I'll go back through and check test coverage.   I'm assuming I'll find unit tests for the different states of SCC, namespace annotations, and constraints.  I think the tests will be easier to read if they focus on one contraint at a time instead of one giant method testing all of them.

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -205,8 +226,11 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
+		requiresNamespaceAllocations := resolveUIDRange || resolveSELinuxLevel || resolveFSGroup || resolveSupplementalGroups
 
-		if resolveUIDRange || resolveSELinuxLevel {
+		if requiresNamespaceAllocations {
```
Below (its not in the diff), don't create a scoped `if` that doesn't scope.

```go
if namespace, err = c.getNamespace(ns, namespace); err != nil {
```
should be 
```go
namespace, err = c.getNamespace(ns, namespace)
if err != nil {
```

The former tricks me into thinking that the variables are scoped and don't leak out, the latter makes it clear that those variables are not scoped to the block.

deads2k commented:
put v1beta3 code into a separate commit so we can squash it into the v1beta3 carry later on.

pweil- commented:
@deads2k - updated for feedback.  Broke into a "normal" and "v1beta3" commit for upstreams.

pweil- commented:
upgrade.md updated.  Current doc that the link is pointing to: https://github.com/openshift/openshift-docs/pull/1078/files#diff-ac08c0344bc91609b46aae5f51152c8d

deads2k commented:
I'll go back through again tomorrow morning, but this is looking good.

deads2k commented on pkg/security/admission/admission.go:
```diff
@@ -205,51 +228,62 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
-				}
+		// Resolve the values from the namespace
+		if resolveUIDRange {
+			if constraint.RunAsUser.UIDRangeMin, constraint.RunAsUser.UIDRangeMax, err = getPreallocatedUIDRange(namespace); err != nil {
```
This isn't scoping the values.  It should be:
```go
constraint.RunAsUser.UIDRangeMin, constraint.RunAsUser.UIDRangeMax, err = getPreallocatedUIDRange(namespace)
if err != nil {
```

pweil- commented:
[test]

soltysh commented:
> the PodSpec was transitively required through the DaemonSetSpec. If your object lives under an object that has conversions here, you probably need it. The really scary question to ask yourself is why they have generated the same conversion function twice? I suspect it was done because they did not properly indicate which schemes/group/version a given group/version relies upon.

I've raised that issue very early when implementing jobs (see https://github.com/kubernetes/kubernetes/issues/12977).

soltysh commented:
@pweil- I've left you some comments, mostly stylistic. Other than that it's looking good.

deads2k commented:
The upstream commits need some renaming before merge.

deads2k commented:
Also, looks like your new test isn't passing:
```
--- FAIL: TestSupplementalGroups-2 (13.37s)
	supplemental_groups_test.go:121: timed out waiting for the condition
FAIL
```

pweil- commented:
@soltysh - comments addressed, please take another pass.   I left the defaulter based on @deads2k comment.

deads2k commented on pkg/security/admission/admission_test.go:
```diff
@@ -137,12 +178,30 @@ func TestAdmit(t *testing.T) {
+		expectedFSGroup   int64
+		expectedSupGroups []int64
+		expectedPriv      bool
+		podLevelOnly      bool
```
godoc on this particular field.

deads2k commented on pkg/security/admission/admission_test.go:
```diff
@@ -205,11 +305,31 @@ func TestAdmit(t *testing.T) {
+
+			if !v.podLevelOnly {
+				// ensure anything we expected to be defaulted on the container level is set
+				if *v.pod.Spec.Containers[0].SecurityContext.RunAsUser != v.expectedUID {
```
Why aren't you be able to predict what the expected values should be?

deads2k commented:
Is there a reasonable spot to actually test that this works end to end?  I'm convinced that this code will make sure the Pod/Containers are marked as expected, but I'd really like a test that makes sure that the expected user/group actually works how we want.  That probably belongs in extended, but it would make me feel a lot better.

deads2k commented:
I added a checklist to the description of the things I'd like addressed before merge.  Nothing too big.

Last three commits lgtm.

pweil- commented:
> Is there a reasonable spot to actually test that this works end to end? I'm convinced that this code will make sure the Pod/Containers are marked as expected, but I'd really like a test that makes sure that the expected user/group actually works how we want. That probably belongs in extended, but it would make me feel a lot better.

The sup group integration test is attempting to ensure that it works all the way down to the docker host config of a launched container.  Support for the fields in docker is 1.8.2 with patches applied.  @pmorie and @swagiaal should be ensuring the docker and beyond portion in addition to the application to the host config so we should be doubly covered on some parts.

pweil- commented:
@pmorie - seems to be failing the int test due to 

```
E1027 10:26:09.747152    1208 mount_linux.go:103] Mount failed: exit status 1
Mounting arguments: tmpfs /tmp/openshift-integration/volume/pods/aade3400-7cb6-11e5-aac7-0ecd0b4a7e19/volumes/kubernetes.io~secret/default-token-nk9tf tmpfs [rootcontext="unconfined_u:object_r:svirt_sandbox_file_t:s0"]
Output: mount: only root can use "--options" option
```
suggestions?

liggitt commented:
you can't run pods in integration... it doesn't start with sudo, so it can't use tmpfs

pweil- commented:
> you can't run pods in integration... it doesn't start with sudo, so it can't use tmpfs

is the router test somehow excluded then?  It starts a router pod.  I thought we used the docker build flag specifically to run things with docker.

pweil- commented:
never mind, it starts it itself I think

pweil- commented:
modified the int test to launch the pod but then not check for the docker host config.  It will now simulate what the kubelet does to craft the host config using the sc provider and checks the modified host config.

deads2k commented:
> you can't run pods in integration... it doesn't start with sudo, so it can't use tmpfs

This should be fixable on jenkins, right?

deads2k commented on test/integration/supplemental_groups_test.go:
```diff
@@ -0,0 +1,108 @@
+		t.Fatalf("unable to get updated pod: %v", err)
+	}
+
+	// simulate what the kubelet would do to modify the container's host config
```
This leaves me unsatisfied.  I'd rather see an extended test get its pod created or fix the integration test invocation to allow running pods.

liggitt replied:
if tmpfs for the SA token is the issue, you can create a volume mount at the expected SA token mount path and the SA token won't get added.

pweil- replied:
I'll give that a shot.  Worse comes to worse I can move this to an extended test.

pweil- commented:
[test][extended:security]

pweil- commented:
@deads2k 

```
test/extended/core.sh --focus="security" -v

=== RUN "test/extended/security/supplemental_groups.go":
... snip ...
STEP: creating a pod that requests supplemental groups
STEP: retrieving the pod and ensuring groups are set
STEP: waiting for the pod to become running
STEP: finding the docker container id on the pod
STEP: creating a docker client and inspecting the container
STEP: ensuring the groups are set
STEP: Destroying namespace "e2e-tests-security-supgroups-qzl16" for this suite.
    ok       security: supplemental groups Ensure supplemental groups propagate to docker should propagate reques ... (9.17716644s)
```

pweil- commented:
re[test][extended]

pweil- commented:
let's try this again without the bad tag....re[test]

pweil- commented:
@pmorie @deads2k @swagiaal - this didn't seem to work in jenkins although it worked fine locally (not run as root as well).  I'm going to add a bit more logging to the test to try and work it out.  Does anyone know the docker version that jenkins runs?  

@swagiaal I recall you saying previously that if the version was too low it would just ignore the parameters.

deads2k commented:
> Does anyone know the docker version that jenkins runs?

@danmcp Do you know offhand?

danmcp commented:
https://ci.openshift.redhat.com/jenkins/job/devenv_ami/lastStableBuild/consoleFull

and search for docker.

docker.x86_64                 1.7.1-115.el7

pweil- commented:
@pmorie @deads2k I added a check on the extended test to check the version before running or skip.  Please take a glance.  New local output is showing the step for ensuring the feature is supported: https://gist.github.com/pweil-/9a783b6f4946e54e4933

pweil- commented:
flake timeout on jenkins in unrelated tests.  re[test]

swagiaal commented:
> @swagiaal I recall you saying previously that if the version was too low it would just ignore the parameters.

@pweil- that is correct

pweil- commented:
it is correctly skipping the test now by checking the docker version:

```
=== RUN "test/extended/security/supplemental_groups.go":
>>> testContext.KubeConfig: /tmp/openshift-extended-tests/core/openshift.local.config/master/admin.kubeconfig
    skip    security: supplemental groups Ensure supplemental groups propagate to docker should propagate reques ... (7.086870098s)
```

It failed on unrelated extended tests.

@pmorie if you want to double check the last three commits to make sure you have the latest I'm fine with closing this out in favor of your updated PR.

Final status here:  

1.  LGTM from @deads2k on the last 3 commits
1.  New extended test will skip unsupported docker versions (clean run output above)
1.  Checklist on the top comment should migrate to @pmorie's PR.  I will create follow ups and he will address other items

soltysh commented:
Add to that my LGTM :+1:

pweil- commented:
Closing in favor of https://github.com/openshift/origin/pull/5169 :birthday:
